### PR TITLE
This fixes #211, fixes #160

### LINKED ISSUE #160: On SQL Error (plv8.execute("erroneous sql")) it's returns Error 0 and SQLState XX000
It seems that on any SQL Error (invalid syntax, invalid column name in select, deadlock, connection dropped, etc) plv8 returns proper error textual description but error code is always 0 and SQLState is XX000.

Additional info: Checked it within Java+SpringBoot+Hibernate project, where on some actions are triggered PostgreSQL SPs written in plv8.

plv8 1.4.4, plv8 1.5.0

### LINKED ISSUE #211: plv8 swallows error stack
If PostgreSQL throws an error up the stack from a `plv8.execute(...)`, the actual error stack is swallowed and hidden. When using `plpgsql`, the actual error line is reported.

Here's an example to demonstrate this:
```sql
CREATE OR REPLACE FUNCTION public._anerror()
  RETURNS boolean AS
$BODY$
BEGIN

  RAISE EXCEPTION 'my error';

  RETURN true;

END;
$BODY$
LANGUAGE plpgsql;

CREATE OR REPLACE FUNCTION public._callsanerror()
  RETURNS boolean AS
$BODY$
DECLARE
  something boolean;
BEGIN
  something := _anerror();

  RETURN something;
END;
$BODY$
  LANGUAGE plpgsql;
```

Then call `_callsanerror()` function either directly or through `plpgsql`:
```sql
SELECT _callsanerror();
```

Outputs the message:
```
ERROR:  my error
CONTEXT:  PL/pgSQL function _callsanerror() line 5 at assignment
********** Error **********

ERROR: my error
SQL state: P0001
Context: PL/pgSQL function _callsanerror() line 5 at assignment
```

```sql
DO $$
BEGIN
  SELECT _callsanerror();
END;
$$ LANGUAGE plpgsql;
```

Outputs the message:
```
ERROR:  my error
CONTEXT:  PL/pgSQL function _callsanerror() line 5 at assignment
SQL statement "SELECT _callsanerror()"
PL/pgSQL function inline_code_block line 3 at SQL statement
********** Error **********

ERROR: my error
SQL state: P0001
Context: PL/pgSQL function _callsanerror() line 5 at assignment
SQL statement "SELECT _callsanerror()"
PL/pgSQL function inline_code_block line 3 at SQL statement
```

When called from `plv8`, the relevate stack line that actually threw the error is swallowed:
```sql
DO $$
  plv8.execute("SELECT _callsanerror();");
$$ LANGUAGE plv8;
```

Outputs the message:
```
ERROR:  my error
DETAIL:  undefined() LINE 2:   plv8.execute("SELECT _callsanerror();");
********** Error **********

ERROR: my error
SQL state: XX000
Detail: undefined() LINE 2:   plv8.execute("SELECT _callsanerror();");
```
No reference to `Context: PL/pgSQL function _callsanerror() line 5 at assignment`.

Is there any way to make use of `PG_EXCEPTION_CONTEXT` and return the full stack?
https://www.postgresql.org/docs/current/static/plpgsql-control-structures.html#PLPGSQL-EXCEPTION-DIAGNOSTICS

### PR DESCRIPTION
**Pull request by bendiy**
Send full error messages: `CODE`, `DETAIL`, `HINT`, `CONTEXT`.
Use `CONTEXT` instead of `DETAIL`.
Return the correct `SQLERRCODE`.
Expose `detail`, `hint`, `context`, `internalquery` and `sqlerrcode` to JavaScript error catching.

Fixes #211 
Fixes #160 
Related to #117 

Raising an exception such as:
```sql
    RAISE EXCEPTION USING
      MESSAGE = 'A First Name is required.',
      DETAIL = 'A detailed message??',
      HINT = 'Do this to avoid that.',
      COLUMN = 'firstName'
      ;
```

The error object in a `try {} catch(err) {}` will now look like:
```json
{
  "code": 16777248,
  "column_name": "firstName",
  "constraint_name": null,
  "context": "PL/pgSQL function _callsanerror() line 5 at assignment\nSQL statement \"SELECT _callsanerror();\"",
  "datatype_name": null,
  "detail": "A detailed message??",
  "hint": "Do this to avoid that.",
  "internalquery": null,
  "message": "A First Name is required.",
  "schema_name": null,
  "sqlerrcode": "P0001",
  "table_name": null
}
```

The error reported from the example in #211:
```sql
DO $$
  plv8.execute("SELECT _callsanerror();");
$$ LANGUAGE plv8;
```
Is now:
```
ERROR:  An error!
DETAIL:  A detailed message??
HINT:  Do this to avoid that.
CONTEXT:  PL/pgSQL function _callsanerror() line 5 at assignment
SQL statement "SELECT _callsanerror();"
undefined() LINE 2:   plv8.execute("SELECT _callsanerror();");
********** Error **********

ERROR: An error!
SQL state: P0001
Detail: A detailed message??
Hint: Do this to avoid that.
Context: PL/pgSQL function _callsanerror() line 5 at assignment
SQL statement "SELECT _callsanerror();"
undefined() LINE 2:   plv8.execute("SELECT _callsanerror();");
```
Matching the `plpgsql` version.

### CODE REVIEW

**Comment by JerrySievert** [comment]
this looks great except that it's not passing against any of the CI tests on travis.  (see https://travis-ci.org/plv8/plv8/jobs/248890124#L3750 for an example).

any chance you can look into that?

thanks!

File: .travis.yml [hunk 1]
```diff
@@ -68,6 +68,10 @@ dist: trusty
 script:
   - make && sudo make install && make installcheck
 
+after_failure:
+  - cat regression.diffs || true
+  - cat regression.out || true
+
 after_script:
```
Comment (JerrySievert) [inline .travis.yml]: you'll note that it does the output by default here, I'd rather keep it outputting regardless, since that includes "failure".

**Comment by bendiy** [comment]
@JerrySievert Yes, I saw something similar on the `master` branch. I think I need to get the right output from your change here:
https://github.com/plv8/plv8/commit/57db50d0dade7a840a8f325e5d0c89ef7fc71466#diff-4e572e5378e9ff017b0e64322304fb73R4

Any ideas on this one?
https://travis-ci.org/plv8/plv8/jobs/248890125#L3829

Appears to be a different output from the different version of V8, so it won't be consistent in the tests.

**Comment by JerrySievert** [comment]
that one is new to me - let me make sure master (which is definitely not a release branch) is up to date with latest release.  give me a few minutes.

**Comment by bendiy** [comment]
My editor strips trailing spaces:
```
CONTEXT:  undefined() LINE 1:  plv8.elog(NOTICE, 'foo = ' + foo)
```
I'll add it back.

**Comment by JerrySievert** [comment]
so does mine - I have to edit those in vim in order for them to work correctly.  the silly things we take for granted.

**Comment by JerrySievert** [comment]
I merged in changes from 2.0.3, if things look good this weekend, I'll cut a release.

note, I'm not planning on back porting this.

**Comment by bendiy** [comment]
Cool.

> note, I'm not planning on back porting this.

That's fine. We're starting work on migrating to the 2.x branch. I really just need this change for development, so that's easy enough to build and use locally.

**Comment by JerrySievert** [comment]
thanks again for the PR, I would have managed to get around to it eventually :)

**Comment by bendiy** [comment]
Last commit should cleanup the docs for pgxn.

**Comment by JerrySievert** [comment]
waiting on travis icon, and my kitten to stop walking on my keyboard

File: plv8.h [hunk 2]
```diff
@@ -48,7 +48,10 @@ class js_error
 {
 private:
 	char	   *m_msg;
+	int	   m_code;
```
Comment (bendiy) [inline plv8.h]: I wasn't sure if this needed to be a pointer, but `errcore()` doesn't take one.

Comment (JerrySievert) [inline plv8.h]: I'll do a close review before release - hence likely a release this weekend.  scary thought, but I actually have a pretty good handle on this stuff, just sometimes a bit slow :)

**Comment by bendiy** [comment]
Passing on the older V8 (4.4.63.31) versions, but still a discrepancy between this test on the newer V8 (5.1.281.14) versions:
https://travis-ci.org/plv8/plv8/jobs/248934351#L3758
```
$ cat regression.diffs || true
*** /home/travis/build/plv8/plv8/expected/plv8.out	2017-06-30 21:41:19.072177299 +0000
--- /home/travis/build/plv8/plv8/results/plv8.out	2017-06-30 22:05:20.907889591 +0000
***************
*** 538,544 ****
  
  -- ERRORS
  CREATE FUNCTION syntax_error() RETURNS text AS '@' LANGUAGE plv8;
! ERROR:  SyntaxError: Unexpected token ILLEGAL
  CONTEXT:  syntax_error() LINE 1: @
  CREATE FUNCTION reference_error() RETURNS text AS 'not_defined' LANGUAGE plv8;
  SELECT reference_error();
--- 538,544 ----
  
  -- ERRORS
  CREATE FUNCTION syntax_error() RETURNS text AS '@' LANGUAGE plv8;
! ERROR:  SyntaxError: Invalid or unexpected token
  CONTEXT:  syntax_error() LINE 1: @
  CREATE FUNCTION reference_error() RETURNS text AS 'not_defined' LANGUAGE plv8;
  SELECT reference_error();
======================================================================
```
I'm not sure there's anything I can do about that.

**Comment by bendiy** [comment]
I didn't even touch `m_msg`, so it has to be something in the V8 dependency tree that changed.
I expect there will be the same CI error in this build:
https://travis-ci.org/plv8/plv8/builds/248935858

**Comment by bendiy** [comment]
If I build `master` locally, when I run:
`CREATE EXTENSION plv8;`
I get:
```
ERROR:  could not load library "/usr/lib/postgresql/9.5/lib/plv8.so": /usr/lib/postgresql/9.5/lib/plv8.so: undefined symbol: _ZTVN2v88internal20SetupIsolateDelegateE
********** Error **********

ERROR: could not load library "/usr/lib/postgresql/9.5/lib/plv8.so": /usr/lib/postgresql/9.5/lib/plv8.so: undefined symbol: _ZTVN2v88internal20SetupIsolateDelegateE
SQL state: XX000
```

**Comment by JerrySievert** [comment]
I'll build locally, but:

* what platform
* did you `make` or `make static`

that's a delegate for `SetupIsolate`, which makes me wonder about `static` or not.

**Comment by JerrySievert** [comment]
sad I can't link
screen shot 2017-06-30 at 3 41 15 pm - /cc @kneemer please help, a link would be amazing! but I'll pull it down and make it work.

**Comment by JerrySievert** [comment]
screen shot 2017-06-30 at 3 41 15 pm

**Comment by bendiy** [comment]
I just merged in your fork's master as recommended here:
https://github.com/plv8/plv8/issues/201

Building with `AUTOV8_VERSION = 5.8.301` now. Ubuntu 16.04 PG 9.5.

**Comment by JerrySievert** [comment]
that's likely the issue.  I'll investigate dropping dependencies for external libraries to better align with node.  not sure the outcome, but I'll at least get a path forward without conflict.

**Comment by bendiy** [comment]
Went ahead and fixed the order for this:
https://github.com/plv8/plv8/issues/194#issuecomment-308421536

**Comment by bendiy** [comment]
Will also _hopefully_ fix: #216

**Comment by bendiy** [comment]
Can close #161

**Comment by bendiy** [comment]
Weird it passed with V8 5.1.281.14:
https://travis-ci.org/plv8/plv8/jobs/248935861
I don't know what would have changed to cause `ERROR:  SyntaxError: Invalid or unexpected token`.

**Comment by JerrySievert** [comment]
last actual release was with `5.4.500.43`, pulling down your branch now to take a look.

**Comment by JerrySievert** [comment]
running now: https://travis-ci.org/JerrySievert/plv8/builds/249195027

updated v8 version for static and for travis to `5.8.301` as well, will see how that looks.  also running a build locally, but that's going to take a while.

**Comment by JerrySievert** [comment]
ok, this is what I'm getting:

```
*** /Users/jerry/work/plv8/expected/plv8.out    2017-07-01 14:18:37.000000000 -0700
--- /Users/jerry/work/plv8/results/plv8.out     2017-07-01 18:11:00.000000000 -0700
***************
*** 538,544 ****

  -- ERRORS
  CREATE FUNCTION syntax_error() RETURNS text AS '@' LANGUAGE plv8;
! ERROR:  SyntaxError: Unexpected token ILLEGAL
  CONTEXT:  syntax_error() LINE 1: @
  CREATE FUNCTION reference_error() RETURNS text AS 'not_defined' LANGUAGE plv8;
  SELECT reference_error();
--- 538,544 ----

  -- ERRORS
  CREATE FUNCTION syntax_error() RETURNS text AS '@' LANGUAGE plv8;
! ERROR:  SyntaxError: Invalid or unexpected token
  CONTEXT:  syntax_error() LINE 1: @
  CREATE FUNCTION reference_error() RETURNS text AS 'not_defined' LANGUAGE plv8;
  SELECT reference_error();

======================================================================
```

this should be a syntax error, but what changed is the error expected.  I think this is safe (I'll make the change locally, and put in a PR to your branch with my changes).

**Comment by JerrySievert** [comment]
oof. this is going to continue into tomorrow.

**Comment by bendiy** [comment]
@JerrySievert Looks good now. Thanks!

**Comment by bendiy** [comment]
I tried a `make static`  of this. It built and installs fine, but I still get the error when running any plv8 code:
```
ERROR:  could not load library "/usr/lib/postgresql/9.5/lib/plv8.so": /usr/lib/postgresql/9.5/lib/plv8.so: undefined symbol: _ZTVN2v88internal20SetupIsolateDelegateE
********** Error **********

ERROR: could not load library "/usr/lib/postgresql/9.5/lib/plv8.so": /usr/lib/postgresql/9.5/lib/plv8.so: undefined symbol: _ZTVN2v88internal20SetupIsolateDelegateE
SQL state: XX000
```

```
dev@dev-desktop-vm:~$ g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/5/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.4.1-2ubuntu1~16.04' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.1 20160904 (Ubuntu 5.4.1-2ubuntu1~16.04)
dev@dev-desktop-vm:~$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/5/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.4.1-2ubuntu1~16.04' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.1 20160904 (Ubuntu 5.4.1-2ubuntu1~16.04)
```
```
"PostgreSQL 9.5.5 on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 5.3.1-14ubuntu2) 5.3.1 20160413, 64-bit"
```

**Comment by JerrySievert** [comment]
odd, I tried it on 17.04, and it worked well.  

it's complaining about `v8::internal::SetupIsolateDelegate` - I'll get a 16.04 vm running with the same version of postgres and give it a compile/test.

**Comment by bendiy** [comment]
I get the same error when building static from the `v2.0.3` tag, but not from `v1.5.7`.

**Comment by JerrySievert** [comment]
that's definitely odd, since I personally use 2.0.3 in production.

ubuntu's installed, have a couple of other things I need to get done this afternoon, but I'll let it compile.

**Comment by bendiy** [comment]
Same for `v2.0.0`.

**Comment by JerrySievert** [comment]
very odd, I'll get to the bottom of it - these aren't new releases, and have been used in production quite a bit.  all of the 2.0.* releases were validated on 16.04 as well.

**Comment by JerrySievert** [comment]
are you sure you don't have a copy of `libv8.so` sitting around somewhere?  that really sounds like an issue that occurs if you are trying to use an old copy of v8.

**Comment by JerrySievert** [comment]
can you do me a favor and run the following:

```
jerry@ubuntu:~$ sudo aptitude search libv8
p   libv8-3.14-dbg                  - V8 JavaScript engine - debugging symbols
p   libv8-3.14-dbg:i386             - V8 JavaScript engine - debugging symbols
p   libv8-3.14-dev                  - V8 JavaScript engine - development files f
p   libv8-3.14-dev:i386             - V8 JavaScript engine - development files f
p   libv8-3.14.5                    - V8 JavaScript engine - runtime library
p   libv8-3.14.5:i386               - V8 JavaScript engine - runtime library
p   libv8-dev                       - V8 JavaScript engine - development files f
p   libv8-dev:i386                  - V8 JavaScript engine - development files f
v   libv8-legacy-dev                -
v   libv8-legacy-dev:i386           -
```

**Comment by bendiy** [comment]
```
dev@dev-desktop-vm:~$ sudo aptitude search libv8
p   libv8-3.14-dbg                                                        - V8 JavaScript engine - debugging symbols
p   libv8-3.14-dbg:i386                                                   - V8 JavaScript engine - debugging symbols
p   libv8-3.14-dev                                                        - V8 JavaScript engine - development files for 3.14 branch
p   libv8-3.14-dev:i386                                                   - V8 JavaScript engine - development files for 3.14 branch
i A libv8-3.14.5                                                          - V8 JavaScript engine - runtime library
p   libv8-3.14.5:i386                                                     - V8 JavaScript engine - runtime library
i   libv8-dev                                                             - V8 JavaScript engine - development files for latest branch
p   libv8-dev:i386                                                        - V8 JavaScript engine - development files for latest branch
v   libv8-legacy-dev                                                      -
v   libv8-legacy-dev:i386                                                 -
```

**Comment by bendiy** [comment]
OK, removed, I'll try a build again.

**Comment by JerrySievert** [comment]
yup, that's going to be your issue.

**Comment by JerrySievert** [comment]
```
jerry@ubuntu:~/plv8$ sudo su postgres
postgres@ubuntu:/home/jerry/plv8$ createdb test
postgres@ubuntu:/home/jerry/plv8$ psql test
psql (9.5.7)
Type "help" for help.

test=# create extension plv8;
CREATE EXTENSION
test=# do $$ return 1; $$ language plv8;
DO
test=# \q
postgres@ubuntu:/home/jerry/plv8$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.2 LTS"
postgres@ubuntu:/home/jerry/plv8$
```

**Comment by bendiy** [comment]
There we go:
```
WARNING:  plv8.version = 2.0.4-dev
```
It's building now. Thanks for the help!

**Comment by bendiy** [comment]
Should be able to close this one now too:
https://github.com/plv8/plv8/issues/167

**Comment by JerrySievert** [comment]
awesome.  now you see why `v8` version `3.14` isn't compatible with modern plv8.  same under the hood changes that node had to deal with.  there was an additional breaking change between plv8 `1.5` and plv8 `2.0`, in how memory is allocated.

**Comment by JerrySievert** [comment]
I'm going to merge this into master - can deal with the outstanding error issues shortly after that.

I'll prep for a 2.1.0 release as well.  going to see if I can move to an even newer v8 before doing that, though, I'd rather be in the 5.8+ range.

After all of that is over, I'll go through the process of getting `PLV8U` up and running, and will open up a discussion about features and API.

**Comment by bendiy** [comment]
Thanks!

**Comment by Saurabh-Bhadane** [comment]
@JerrySievert  I am using plv8 version 2.1.0 but unable to get a complete error object as mentioned previously.
expected  :  {
{
  "code": 16777248,
  "column_name": "firstName",
  "constraint_name": null,
  "context": "PL/pgSQL function _callsanerror() line 5 at assignment\nSQL statement \"SELECT _callsanerror();\"",
  "datatype_name": null,
  "detail": "A detailed message??",
  "hint": "Do this to avoid that.",
  "internalquery": null,
  "message": "A First Name is required.",
  "schema_name": null,
  "sqlerrcode": "P0001",
  "table_name": null
}

but getting same plv8 exception output as mentioned on #211 .
Please assist

**Pull request merged**